pkg/operator/controller.go

```diff
@@ -1,484 +1,331 @@
-	if err != nil {
-		return nil, err
+			return !reflect.DeepEqual(profileOld.Status.Conditions, profileNew.Status.Conditions)
+		},
 	}
 
-	listers := &ntoclient.Listers{}
-	clients := &ntoclient.Clients{}
-	controller := &Controller{
-		kubeconfig: kubeconfig,
-		workqueue:  workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		listers:    listers,
-		clients:    clients,
-		pc:         NewProfileCalculator(listers, clients),
+	podPredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return r.state.enablePodEvents
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return r.state.enablePodEvents
+		},
 	}
 
-	// Initial event to bootstrap CR if it doesn't exist.
-	controller.workqueue.AddRateLimited(wqKey{kind: wqKindTuned, name: tunedv1.TunedDefaultResourceName})
+	// In order to use cached client list filtering for pods: client.MatchingFields{"spec.nodeName": mNodeName}, there is a need to use and index field.
+	// The alternative is to use an uncached client.
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &corev1.Pod{}, "spec.nodeName", func(o client.Object) []string {
+		pod := o.(*corev1.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
 
-	controller.clients.Kube, err = kubeset.NewForConfig(controller.kubeconfig)
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&tunedv1.Tuned{}).
+		Owns(&tunedv1.Profile{}, builder.WithPredicates(profilePredicates)).
+		Watches(
+			&source.Kind{Type: &corev1.Node{}},
+			handler.EnqueueRequestsFromMapFunc(r.nodeToTunedRequests),
+			//TODO - consider using builder.OnlyMetadata instead these perdicats
+			builder.WithPredicates(predicate.LabelChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &corev1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(r.podLabelsToTuned),
+			builder.WithPredicates(podPredicates,
+				predicate.And(
+					predicate.Funcs{UpdateFunc: func(e event.UpdateEvent) bool {
+						return r.state.enablePodEvents
+					}}, predicate.LabelChangedPredicate{}))).
+		Watches(
+			&source.Kind{Type: &mcov1.MachineConfigPool{}},
+			handler.EnqueueRequestsFromMapFunc(r.mcpToTunedRequests)).
+		Complete(r)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	r.state.enablePodEvents = false
+	return nil
+}
 
-	// ClusterOperator
-	controller.clients.ConfigV1Client, err = configv1client.NewForConfig(controller.kubeconfig)
+func (r *TunedReconciler) nodeToTunedRequests(nodeObj client.Object) []reconcile.Request {
+	node := nodeObj.(*corev1.Node)
+	if node.DeletionTimestamp != nil {
+		profile := &tunedv1.Profile{}
+		key := types.NamespacedName{
+			Namespace: ntoconfig.OperatorNamespace(),
+			Name:      node.Name,
+		}
+		if err := r.Get(context.TODO(), key, profile); err != nil {
+			klog.Errorf("could not get profile for node %s deletion event %v", node.Name, err)
+			return nil
+		}
+		if err := r.Delete(context.TODO(), profile); err != nil {
```

Comment:
I remember we discussed it, WDT about moving deletion logic to reconcile loop?
I think one of the reasons why you should not delete the object here is because some of the commands can fail and you do not have control to re-enter the reconcile loop.

Reply:
will be added in the next push